### PR TITLE
BUILD/UCT/IB: added configuration to disable multiple resource domains

### PR DESCRIPTION
**Pull request by hoopoepg**

### CODE REVIEW

**Comment by yosefe** [comment]
we cannot really use mlx5 transports without resource domain, otherwise we could have data corruption on the BF register if it's shared between verbs and mlx5

**Comment by hoopoepg** [comment]
right, but it at least allows to identify issue faster

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/7395/ for details (Mellanox internal link).

**Comment by rzambre** [comment]
@yosefe The corruption on the BF register could occur if the user creates an `rc` iface and an `rc_mlx5` iface within the same memory domain. But if the user uses separate memory domains for the two ifaces, then data corruption would not be an issue. Is my understanding correct?

File: config/m4/ib.m4 [hunk 1]
```diff
@@ -7,6 +7,11 @@
 #
 
 
+AC_ARG_ENABLE(res-domain,
```
Comment (yosefe) [inline config/m4/ib.m4]: no need for changing ib.m4, the config can be runtime only

File: src/uct/ib/base/ib_iface.c [hunk 2]
```diff
@@ -536,7 +542,11 @@ static int uct_ib_iface_res_domain_cmp(uct_ib_iface_res_domain_t *res_domain,
 #if HAVE_IBV_EXP_RES_DOMAIN
     uct_ib_device_t *dev = uct_ib_iface_device(iface);
 
-    return res_domain->ibv_domain->context == dev->ibv_context;
+    if (iface->config.res_domain) {
```
Comment (yosefe) [inline src/uct/ib/base/ib_iface.c]: return !iface->config.res_domain || (res_domain->ibv_domain->context == dev->ibv_context)

File: src/uct/ib/base/ib_iface.c [hunk 3]
```diff
@@ -550,43 +560,47 @@ uct_ib_iface_res_domain_init(uct_ib_iface_res_domain_t *res_domain,
...
-        attr.thread_model = IBV_EXP_THREAD_SINGLE;
-        break;
-    case UCS_THREAD_MODE_SERIALIZED:
-        attr.thread_model = IBV_EXP_THREAD_UNSAFE;
-        break;
-    default:
-        attr.thread_model = IBV_EXP_THREAD_SAFE;
-        break;
-    }
+    if (iface->config.res_domain) {
```
Comment (yosefe) [inline src/uct/ib/base/ib_iface.c]: instead, we can call uct_worker_tl_data_get() only if res_domain==1, in ib_iface.c:667
```
if (res_domain_key == UCT_IB_IFACE_NULL_RES_DOMAIN_KEY || !res_domain) {
  self->res_domain = NULL;
} else {
...
```

File: src/uct/ib/base/ib_iface.c [hunk 4]
```diff
@@ -744,13 +760,14 @@ UCS_CLASS_INIT_FUNC(uct_ib_iface_t, uct_ib_iface_ops_t *ops, uct_md_h md,
...
+    if (self->config.res_domain) {
+        ibv_destroy_cq(self->cq[UCT_IB_DIR_TX]);
+    }
 err_destroy_comp_channel:
     ibv_destroy_comp_channel(self->comp_channel);
 err_put_res_domain:
-    if (self->res_domain != NULL) {
-        uct_worker_tl_data_put(self->res_domain, uct_ib_iface_res_domain_cleanup);
-    }
+    uct_worker_tl_data_put(self->res_domain, uct_ib_iface_res_domain_cleanup,
```
Comment (yosefe) [inline src/uct/ib/base/ib_iface.c]: ` if (self->res_domain != NULL) {` should suffice here

File: src/uct/ib/base/ib_iface.h [hunk 5]
```diff
@@ -145,6 +148,7 @@ struct uct_ib_iface {
         uint8_t             sl;
         uint8_t             traffic_class;
         size_t              max_iov;             /* Maximum buffers in IOV array */
+        int                 res_domain;          /* Enable multiple resource domains */
```
Comment (yosefe) [inline src/uct/ib/base/ib_iface.h]: can move this var before max_iov to reduce overall struct size

File: src/uct/ib/rc/base/rc_iface.c [hunk 6]
```diff
@@ -952,7 +952,7 @@ ucs_status_t uct_rc_iface_qp_create(uct_rc_iface_t *iface, int qp_type,
 #  endif
 
 #if HAVE_IBV_EXP_RES_DOMAIN
-    if (iface->super.res_domain != NULL) {
+    if (iface->super.config.res_domain && (iface->super.res_domain != NULL)) {
```
Comment (yosefe) [inline src/uct/ib/rc/base/rc_iface.c]: should be enough to check `iface->super.res_domain != NULL`

File: config/m4/ib.m4 (hunk 1 above)
Comment (hoopoepg) [reply config/m4/ib.m4]: removed

File: src/uct/ib/base/ib_iface.c (hunk 2 above)
Comment (hoopoepg) [reply src/uct/ib/base/ib_iface.c]: no more actual

File: src/uct/ib/base/ib_iface.c (hunk 3 above)
Comment (hoopoepg) [reply src/uct/ib/base/ib_iface.c]: done

File: src/uct/ib/base/ib_iface.h (hunk 5 above)
Comment (hoopoepg) [reply src/uct/ib/base/ib_iface.h]: moved

File: src/uct/ib/rc/base/rc_iface.c (hunk 6 above)
Comment (hoopoepg) [reply src/uct/ib/rc/base/rc_iface.c]: fixed

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/7416/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/7417/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/7418/ for details (Mellanox internal link).

**Comment by hoopoepg** [comment]
@yosefe I changes config semantic: for now it disables multiple domains (instead of enable it by default)

File: src/uct/ib/base/ib_iface.c [hunk 7]
```diff
@@ -150,6 +150,12 @@ ucs_config_field_t uct_ib_iface_config_table[] = {
    "Which pkey value to use. Should be between 0 and 0x7fff.",
    ucs_offsetof(uct_ib_iface_config_t, pkey_value), UCS_CONFIG_TYPE_HEX},
 
+#if HAVE_IBV_EXP_RES_DOMAIN
+  {"DISABLE_RESOURCE_DOMAIN", "n",
```
Comment (yosefe) [inline src/uct/ib/base/ib_iface.c]: let's make it positive e.g "UCX_RESOURCE_DOMAIN=y"
to be same spirit as other parameters

Comment (hoopoepg) [reply src/uct/ib/base/ib_iface.c]: done

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/7419/ for details (Mellanox internal link).

**Comment by hoopoepg** [comment]
@yosefe ok to merge?

**Comment by rzambre** [comment]
With the merged PR, is there a way to disable resource domains for rc_mlx5 transport?

**Comment by hoopoepg** [comment]
@rzambre yes, set UCX_MEM_RESOURCE_DOMAIN=n to disable it

**Comment by rzambre** [comment]
@hoopoepg I get
> UCX  WARN  unused env variable: UCX_MEM_RESOURCE_DOMAIN

I think you meant UCX_RC_MLX5_RESOURCE_DOMAIN=n :)

**Comment by hoopoepg** [comment]
@rzambre sorry, my fault
you are right, it will work, or more general variable: UCX_IB_RESOURCE_DOMAIN=n

**Comment by rzambre** [comment]
Aah, got it 👍

**Pull request merged**